pytensor/link/jax/dispatch/tensor_basic.py

```diff
@@ -193,3 +217,25 @@ def scalar_from_tensor(x):
         return jnp.array(x).flatten()[0]
 
     return scalar_from_tensor
+
+
+@jax_funcify.register(Tri)
+def jax_funcify_Tri(op, node, **kwargs):
+    dtype = op.dtype
+    tri_args = node.inputs
+    constant_args = []
+
+    for arg in tri_args:
+        if not isinstance(arg, Constant):
```

Comment:
`if isinstance(arg, TensorVariable)` instead?

Reply:
You should use the inputs passed to the inner function

Reply:
Maybe the most flexible solution is to use constants when we have but not fail if we don't have them:

```
@jax_funcify.register(Tri)
def jax_funcify_Tri(op, node, **kwargs):
    dtype = op.dtype
    N, M, k = node.inputs

    constant_N = None if not isinstance(N, Constant) else N.data
    constant_M = None if not isinstance(M, Constant) else M.data
    constant_k = None if not isinstance(k, Constant) else k.data

    def tri(N, M, k):
        if constant_N is not None:
          N = constant_N
        ... # same for others
        return jnp.tri(N, M, k, dtype=dtype)

    return tri
```

Because we don't know if PyTensor non-constants will look like constants to JAX when it is being jitted. The common case has to do with shapes of inputs which are always constant as far as JAX is concerned, but not PyTensor.

Reply:
The simplest implementation works in this case:

```python
    out = at.tri(10, 10, 0)
```

But not this case:

```
    M = scalar('M', dtype='int64')
    N = scalar('N', dtype='int64')
    k = scalar('k', dtype='int64')
    out = at.tri(M, N, k)
```

I guess the second case is always going to fail? I thought it would work with the is/else version, but I think I'm not understanding something.

Reply:
The second case will always fail I think, not even jax allows that. What can work is something that depends on the input shape

Reply:
OK that's what I thought. I was confused because I was thinking the inputs to the inner `tri` function would somehow magically end up being specified before the graph was compiled to JAX because of the `N = constant_N if constant_N is not None else N` lines in the inner function.

So what should happen if a user tries to do something like the 2nd case? Do you want the error message to raise, or just to fail loudly with the expection JAX produces?

Reply:
Just fail with whatever JAX does. That's exactly the same error if a user tries:

```python
import jax

@jax.jit
def foo(M, N, K):
  return jax.numpy.tri(M, N, K)

foo(5, 5, 0)
```

Reply:
Got it. I'll update the PR when I get home tonight.